From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bastiaan van der Plaat <bastiaan.v.d.plaat@gmail.com>
Date: Mon, 12 Oct 2026 07:54:52 +0200
Subject: [PATCH] Add DS sound

The sounds load when they first play instead of all with the map, within
a budget of main RAM, 768 KiB on a DS, the least recently played ones
that don't play are freed for new ones, also when the main RAM runs
short, and all of them when the next map loads. They are 8-bit and play
on the hardware channels, nothing is mixed.
---
 client/snd_dma.c |  33 +++---
 client/snd_loc.h |  11 +-
 client/snd_mem.c |  71 ++++++++++--
 nds/snd_nds.c    | 293 +++++++++++++++++++++++++++++++++++++++++++++++
 4 files changed, 379 insertions(+), 29 deletions(-)
 create mode 100644 nds/snd_nds.c

diff --git a/client/snd_dma.c b/client/snd_dma.c
index 09e408d..2332c04 100644
--- a/client/snd_dma.c
+++ b/client/snd_dma.c
@@ -183,8 +183,7 @@ void S_Shutdown(void)
 	{
 		if (!sfx->name[0])
 			continue;
-		if (sfx->cache)
-			Z_Free (sfx->cache);
+		S_FreeSound (sfx);
 		memset (sfx, 0, sizeof(*sfx));
 	}
 
@@ -292,8 +291,16 @@ S_BeginRegistration
 */
 void S_BeginRegistration (void)
 {
+	int		i;
+
 	s_registration_sequence++;
 	s_registering = true;
+
+	// the sounds of the last map make room for the loading of the next, they
+	// load again when they play
+	S_StopAllSounds ();
+	for (i = 0; i < num_sfx; i++)
+		S_FreeSound (&known_sfx[i]);
 }
 
 /*
@@ -338,8 +345,7 @@ void S_EndRegistration (void)
 			continue;
 		if (sfx->registration_sequence != s_registration_sequence)
 		{	// don't need this sound
-			if (sfx->cache)	// it is possible to have a leftover
-				Z_Free (sfx->cache);	// from a server that didn't finish loading
+			S_FreeSound (sfx);	// it is possible to have a leftover
 			memset (sfx, 0, sizeof(*sfx));
 		}
 		else
@@ -353,14 +359,7 @@ void S_EndRegistration (void)
 
 	}
 
-	// load everything in
-	for (i=0, sfx=known_sfx ; i < num_sfx ; i++,sfx++)
-	{
-		if (!sfx->name[0])
-			continue;
-		S_LoadSound (sfx);
-	}
-
+	// the sounds load when they first play, they don't all fit
 	s_registering = false;
 }
 
@@ -578,7 +577,10 @@ void S_IssuePlaysound (playsound_t *ps)
 
 	ch->pos = 0;
 	sc = S_LoadSound (ch->sfx);
-    ch->end = paintedtime + sc->length;
+	if (sc)
+		ch->end = paintedtime + sc->length;
+	else
+		ch->sfx = NULL;
 
 	// free the playsound
 	S_FreePlaysound (ps);
@@ -852,7 +854,7 @@ void S_AddLoopSounds (void)
 		sfx = cl.sound_precache[sounds[i]];
 		if (!sfx)
 			continue;		// bad sound effect
-		sc = sfx->cache;
+		sc = S_LoadSound (sfx);
 		if (!sc)
 			continue;
 
@@ -1119,9 +1121,6 @@ void S_Update_(void)
 
 	SNDDMA_BeginPainting ();
 
-	if (!dma.buffer)
-		return;
-
 // Updates DMA time
 	GetSoundtime();
 
diff --git a/client/snd_loc.h b/client/snd_loc.h
index 6208c43..b4a57b7 100644
--- a/client/snd_loc.h
+++ b/client/snd_loc.h
@@ -30,6 +30,7 @@ typedef struct
 {
 	int 		length;
 	int 		loopstart;
+	int			size;			// bytes, with this header
 	int 		speed;			// not needed, because converted on load?
 	int 		width;
 	int 		stereo;
@@ -42,6 +43,7 @@ typedef struct sfx_s
 	int			registration_sequence;
 	sfxcache_t	*cache;
 	char 		*truename;
+	int			lastused;		// paintedtime it last played
 } sfx_t;
 
 // a playsound_t will be generated by each call to S_StartSound,
@@ -135,7 +137,7 @@ extern	vec3_t	listener_up;
 extern	dma_t	dma;
 extern	playsound_t	s_pendingplays;
 
-#define	MAX_RAW_SAMPLES	8192
+#define	MAX_RAW_SAMPLES	1		// the cinematics play without sound
 extern	portable_samplepair_t	s_rawsamples[MAX_RAW_SAMPLES];
 
 extern cvar_t	*s_volume;
@@ -153,6 +155,13 @@ void S_InitScaletable (void);
 
 sfxcache_t *S_LoadSound (sfx_t *s);
 
+// The loaded sounds take at most snd_cachemax bytes, the least recently
+// played are freed for new ones, unless they are still playing
+extern	int		snd_cachemax;
+void S_FreeSound (sfx_t *s);
+qboolean SNDDMA_IsPlaying (sfxcache_t *sc);
+void SNDDMA_SoundFreed (sfxcache_t *sc);
+
 void S_IssuePlaysound (playsound_t *ps);
 
 void S_PaintChannels(int endtime);
diff --git a/client/snd_mem.c b/client/snd_mem.c
index 7dd8d9c..0becde0 100644
--- a/client/snd_mem.c
+++ b/client/snd_mem.c
@@ -21,6 +21,7 @@ Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 
 #include "client.h"
 #include "snd_loc.h"
+#include <malloc.h>
 
 int			cache_full_cycle;
 
@@ -76,8 +77,8 @@ void ResampleSfx (sfx_t *sfx, int inrate, int inwidth, byte *data)
 		{
 			srcsample = samplefrac >> 8;
 			samplefrac += fracstep;
-			if (inwidth == 2)
-				sample = LittleShort ( ((short *)data)[srcsample] );
+			if (inwidth == 2)	// the DS is little endian like the files
+				sample = ((short *)data)[srcsample];
 			else
 				sample = (int)( (unsigned char)(data[srcsample]) - 128) << 8;
 			if (sc->width == 2)
@@ -90,6 +91,38 @@ void ResampleSfx (sfx_t *sfx, int inrate, int inwidth, byte *data)
 
 //=============================================================================
 
+int		snd_cachemax = 0x7fffffff;
+static int	snd_cachebytes;
+
+void S_FreeSound (sfx_t *s)
+{
+	if (!s->cache)
+		return;
+	SNDDMA_SoundFreed (s->cache);
+	snd_cachebytes -= s->cache->size;
+	free (s->cache);
+	s->cache = NULL;
+}
+
+// Frees the least recently played sound that doesn't play, false when there
+// is none
+static qboolean S_FreeOldest (sfx_t *loading)
+{
+	extern sfx_t	known_sfx[];
+	extern int		num_sfx;
+	sfx_t			*sfx, *oldest = NULL;
+	int				i;
+
+	for (i = 0, sfx = known_sfx; i < num_sfx; i++, sfx++)
+		if (sfx->cache && sfx != loading && !SNDDMA_IsPlaying (sfx->cache)
+			&& (!oldest || sfx->lastused < oldest->lastused))
+			oldest = sfx;
+	if (!oldest)
+		return false;
+	S_FreeSound (oldest);
+	return true;
+}
+
 /*
 ==============
 S_LoadSound
@@ -105,16 +138,17 @@ sfxcache_t *S_LoadSound (sfx_t *s)
 	sfxcache_t	*sc;
 	int		size;
 	char	*name;
+	FILE	*f;
 
 	if (s->name[0] == '*')
 		return NULL;
 
 // see if still in memory
+	s->lastused = paintedtime;
 	sc = s->cache;
 	if (sc)
 		return sc;
 
-//Com_Printf ("S_LoadSound: %x\n", (int)stackbuf);
 // load it in
 	if (s->truename)
 		name = s->truename;
@@ -128,33 +162,48 @@ sfxcache_t *S_LoadSound (sfx_t *s)
 
 //	Com_Printf ("loading %s\n",namebuffer);
 
-	size = FS_LoadFile (namebuffer, (void **)&data);
-
-	if (!data)
+	// the file only while it's converted, in the far heap
+	size = FS_FOpenFile (namebuffer, &f);
+	if (size < 0 || !(data = far_malloc (size)))
 	{
+		if (size >= 0)
+			FS_FCloseFile (f);
 		Com_DPrintf ("Couldn't load %s\n", namebuffer);
 		return NULL;
 	}
+	FS_Read (data, size, f);
+	FS_FCloseFile (f);
 
 	info = GetWavinfo (s->name, data, size);
 	if (info.channels != 1)
 	{
 		Com_Printf ("%s is a stereo sample\n",s->name);
-		FS_FreeFile (data);
+		free (data);
 		return NULL;
 	}
 
 	stepscale = (float)info.rate / dma.speed;	
 	len = info.samples / stepscale;
 
-	len = len * info.width * info.channels;
+	len = len * (s_loadas8bit->value ? 1 : info.width) * info.channels;
 
-	sc = s->cache = Z_Malloc (len + sizeof(sfxcache_t));
+	// the least recently played sounds make room, also when there is too
+	// little main RAM left
+	size = len + sizeof(sfxcache_t);
+	while (snd_cachebytes + size > snd_cachemax && S_FreeOldest (s))
+		;
+	while (!(sc = malloc (size)) && S_FreeOldest (s))
+		;
 	if (!sc)
 	{
-		FS_FreeFile (data);
+		free (data);
+		Com_DPrintf ("Couldn't load %s\n", namebuffer);
 		return NULL;
 	}
+	memset (sc, 0, size);
+	s->cache = sc;
+	snd_cachebytes += size;
+	sc->size = size;
 	
 	sc->length = info.samples;
 	sc->loopstart = info.loopstart;
@@ -164,7 +213,7 @@ sfxcache_t *S_LoadSound (sfx_t *s)
 
 	ResampleSfx (s, sc->speed, sc->width, data + info.dataofs);
 
-	FS_FreeFile (data);
+	free (data);
 
 	return sc;
 }
diff --git a/nds/snd_nds.c b/nds/snd_nds.c
new file mode 100644
index 0000000..5778a5c
--- /dev/null
+++ b/nds/snd_nds.c
@@ -0,0 +1,293 @@
+/*
+Copyright (C) 1997-2001 Id Software, Inc.
+
+This program is free software; you can redistribute it and/or
+modify it under the terms of the GNU General Public License
+as published by the Free Software Foundation; either version 2
+of the License, or (at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+
+See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+
+*/
+// snd_nds.c -- Nintendo DS sound output
+//
+// Quake II keeps its channels, which sound plays where at what volume, but
+// instead of mixing them in software the loudest 16 play on the hardware
+// channels, which do the resampling, volume and stereo panning. The samples
+// play straight from the loaded sounds, 8-bit signed, at 11025 Hz on a DS
+// and 22050 Hz on a DSi, which has the RAM for them. The sound hardware
+// can't read the Memory Expansion Pak, they are in main RAM. The play
+// position Quake II's timing needs follows from the tick timer.
+
+#include "../client/client.h"
+#include "../client/snd_loc.h"
+
+#include <nds.h>
+
+#define	MIN(a,b)		((a) < (b) ? (a) : (b))
+#define	CLAMP(lo,x,hi)	((x) < (lo) ? (lo) : (x) > (hi) ? (hi) : (x))
+
+#define	SND_FRAMES		4096		// of the time base, a power of two
+#define	HW_CHANNELS		16
+
+// What the loaded sounds may take
+#define	SND_CACHE_DS	(768 * 1024)
+#define	SND_CACHE_DSI	(3 * 1024 * 1024)
+
+typedef struct
+{
+	channel_t	*ch;		// the Quake II channel playing, NULL for none
+	sfx_t		*sfx;		// a looping sound of entities keeps its channel by this
+	sfxcache_t	*sc;
+	qboolean	autosound;
+	unsigned	vol, pan;
+} hwchannel_t;
+
+extern int			soundtime;
+
+static hwchannel_t	hw[HW_CHANNELS];
+static unsigned	snd_timer;
+static u64		snd_start;
+
+static void HW_Stop (int slot)
+{
+	if (hw[slot].sc)
+		soundStop (BIT(slot));
+	memset (&hw[slot], 0, sizeof(hw[slot]));
+}
+
+qboolean SNDDMA_IsPlaying (sfxcache_t *sc)
+{
+	int	slot;
+
+	for (slot = 0; slot < HW_CHANNELS; slot++)
+		if (hw[slot].sc == sc)
+			return true;
+	return false;
+}
+
+// A sound that is freed stops, the channel would play what comes there next
+void SNDDMA_SoundFreed (sfxcache_t *sc)
+{
+	int	slot;
+
+	for (slot = 0; slot < HW_CHANNELS; slot++)
+		if (hw[slot].sc == sc)
+			HW_Stop (slot);
+}
+
+qboolean SNDDMA_Init (void)
+{
+	dma.channels = 2;
+	dma.samplebits = 16;
+	dma.speed = isDSiMode () ? 22050 : 11025;
+	dma.samples = SND_FRAMES * 2;
+	dma.samplepos = 0;
+	dma.submission_chunk = 1;
+	dma.buffer = NULL;		// nothing is mixed, the samples play as they are
+	Cvar_Set ("s_loadas8bit", "1");
+	snd_cachemax = isDSiMode () ? SND_CACHE_DSI : SND_CACHE_DS;
+
+	soundInit ();
+	soundPowerOn ();
+	soundSetMixerVolume (127);
+	memset (hw, 0, sizeof(hw));
+	snd_timer = soundTimerFromHz (dma.speed);
+	snd_start = tickGetCount ();
+	return true;
+}
+
+int SNDDMA_GetDMAPos (void)
+{
+	// Ticks run at SYSTEM_CLOCK / 64, a sample lasts snd_timer SYSTEM_CLOCK / 2 cycles
+	u64	frames = (tickGetCount () - snd_start) * 32 / snd_timer;
+
+	dma.samplepos = (frames & (SND_FRAMES - 1)) * 2;
+	return dma.samplepos;
+}
+
+void SNDDMA_Shutdown (void)
+{
+	int	i;
+
+	for (i = 0; i < HW_CHANNELS; i++)
+		HW_Stop (i);
+}
+
+void SNDDMA_BeginPainting (void)
+{
+}
+
+void SNDDMA_Submit (void)
+{
+}
+
+void S_InitScaletable (void)
+{
+}
+
+// The volume cvar as 0..256, worked out once for each S_PaintChannels
+static unsigned	master_volume;
+
+// The volume (0..2047) and pan (0..127) of a channel, with the volume cvar
+static void HW_Volume (const channel_t *ch, unsigned *vol, unsigned *pan)
+{
+	int	left = MIN (ch->leftvol, 255), right = MIN (ch->rightvol, 255);
+
+	*vol = MIN ((left + right) * 4 * master_volume >> 8, 2047);
+	*pan = left + right ? right * 127 / (left + right) : 64;
+}
+
+// Starts the sound of a channel on a hardware channel
+static void HW_Start (int slot, channel_t *ch, sfxcache_t *sc)
+{
+	hwchannel_t	*h = &hw[slot];
+	uintptr_t	start = (uintptr_t) sc->data & ~3;	// whole words from an aligned address
+	unsigned	bytes = sc->length * sc->width, loop;
+	qboolean	loops = ch->autosound || sc->loopstart >= 0;
+	// a loop ends at the last whole word, the bytes after the data aren't
+	// part of the sound
+	unsigned	words = loops ? bytes / 4 : (bytes + 3) / 4;
+
+	h->ch = ch;
+	h->sfx = ch->sfx;
+	h->sc = sc;
+	h->autosound = ch->autosound;
+	HW_Volume (ch, &h->vol, &h->pan);
+	// the ARM7 plays from main RAM, past the data cache
+	DC_FlushRange ((void *) start, words * 4);
+	if (loops)
+	{
+		loop = ch->autosound ? 0 : sc->loopstart * sc->width / 4;
+		soundPreparePcm (slot | SOUND_START, h->vol, h->pan, soundTimerFromHz (sc->speed), SoundMode_Repeat,
+			sc->width == 2 ? SoundFmt_Pcm16 : SoundFmt_Pcm8, (const void *) start, loop, words - loop);
+	}
+	else
+		soundPreparePcm (slot | SOUND_START, h->vol, h->pan, soundTimerFromHz (sc->speed), SoundMode_OneShot,
+			sc->width == 2 ? SoundFmt_Pcm16 : SoundFmt_Pcm8, (const void *) start, 0, words);
+	// started, a new sound on the channel sets it back to 0
+	ch->pos = 1;
+}
+
+// Whether a hardware channel plays what a Quake II channel plays
+static qboolean HW_Plays (hwchannel_t *h, channel_t *ch)
+{
+	if (!h->sc || !ch->sfx || h->autosound != ch->autosound)
+		return false;
+	// the looping sounds of entities are made again every frame
+	if (ch->autosound)
+		return h->sfx == ch->sfx;
+	return h->ch == ch && h->sfx == ch->sfx && ch->pos;
+}
+
+/*
+================
+S_PaintChannels
+
+Instead of mixing up to endtime, keeps the hardware channels playing what
+the loudest Quake II channels play
+================
+*/
+void S_PaintChannels (int endtime)
+{
+	channel_t	*ch, *loudest[HW_CHANNELS];
+	playsound_t	*ps;
+	sfxcache_t	*sc;
+	unsigned	vol, pan, loudvol[HW_CHANNELS];
+	int			i, j, k, slot;
+
+	master_volume = CLAMP (0, s_volume->value, 1) * 256;
+
+	// start the playsounds that are due
+	while ((ps = s_pendingplays.next) != &s_pendingplays && ps->begin <= soundtime)
+		S_IssuePlaysound (ps);
+
+	// sounds that ended stop, looping ones go on, those whose samples were
+	// freed for others are over
+	for (i = 0, ch = channels; i < MAX_CHANNELS; i++, ch++)
+	{
+		if (!ch->sfx || ch->autosound)
+			continue;
+		if (!(sc = ch->sfx->cache))
+		{
+			ch->sfx = NULL;
+			continue;
+		}
+		while (ch->end <= soundtime)
+		{
+			if (sc->loopstart < 0)
+			{
+				ch->sfx = NULL;
+				break;
+			}
+			ch->end += sc->length - sc->loopstart;
+		}
+	}
+
+	// the loudest ones, which stay loaded
+	memset (loudest, 0, sizeof(loudest));
+	memset (loudvol, 0, sizeof(loudvol));
+	for (i = 0, ch = channels; i < MAX_CHANNELS; i++, ch++)
+	{
+		if (!ch->sfx || !ch->sfx->cache)
+			continue;
+		HW_Volume (ch, &vol, &pan);
+		if (!vol)
+			continue;
+		for (j = 0; j < HW_CHANNELS && vol <= loudvol[j]; j++)
+			;
+		if (j == HW_CHANNELS)
+			continue;
+		for (k = HW_CHANNELS - 1; k > j; k--)
+		{
+			loudest[k] = loudest[k - 1];
+			loudvol[k] = loudvol[k - 1];
+		}
+		loudest[j] = ch;
+		loudvol[j] = vol;
+	}
+	for (i = 0; i < HW_CHANNELS; i++)
+		if (loudest[i])
+			loudest[i]->sfx->lastused = paintedtime;
+
+	// those still playing keep their hardware channel, the sounds that ended
+	// or are too quiet now stop
+	for (slot = 0; slot < HW_CHANNELS; slot++)
+	{
+		for (i = 0; i < HW_CHANNELS && !(loudest[i] && HW_Plays (&hw[slot], loudest[i])); i++)
+			;
+		if (i == HW_CHANNELS)
+		{
+			if (hw[slot].sc)
+				HW_Stop (slot);
+			continue;
+		}
+		hw[slot].ch = loudest[i];
+		HW_Volume (loudest[i], &vol, &pan);
+		if (vol != hw[slot].vol)
+			soundChSetVolume (slot, hw[slot].vol = vol);
+		if (pan != hw[slot].pan)
+			soundChSetPan (slot, hw[slot].pan = pan);
+		loudest[i] = NULL;
+	}
+
+	// and the new ones start
+	for (i = 0, slot = 0; i < HW_CHANNELS; i++)
+	{
+		if (!loudest[i])
+			continue;
+		while (hw[slot].sc)
+			slot++;
+		HW_Start (slot, loudest[i], loudest[i]->sfx->cache);
+	}
+
+	paintedtime = endtime;
+}
